File: clients/matrix-java/src/main/java/io/ioka/munarium/matrix/MatrixClient.java
// SPDX-License-Identifier: Apache-2.0
package io.ioka.munarium.matrix;

import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.JsonNode;
import java.io.IOException;
import java.net.URI;
import java.net.URLEncoder;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpRequest.BodyPublisher;
import java.net.http.HttpRequest.BodyPublishers;
import java.net.http.HttpResponse;
import java.net.http.HttpResponse.BodyHandlers;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

/**
 * The official Java client for <b>Munarium Matrix</b>, the structured-evidence
 * plane. Synchronous; {@link AsyncMatrixClient} is the {@code CompletableFuture}
 * twin over the same code.
 *
 * <p>It speaks Matrix's REST API and nothing else. <b>There is no gRPC
 * transport here.</b> Matrix's gRPC plane serves {@code MatrixQuery/Execute}
 * alone, and that call is service-to-service: the munarium-server makes it
 * while answering a turn, carrying a session's authorization snapshot an
 * application does not hold. When Matrix grows an RPC an application is
 * entitled to make, this class grows a transport rather than acquiring a
 * sibling.
 *
 * <p>Three absences are design decisions, not gaps:
 *
 * <ul>
 *   <li><b>No sealing.</b> No method on this class seals evidence. A manifest
 *       is a statement about work the <i>sealer</i> did; an SDK offering
 *       {@code sealEvidence} would invite an application to assert provenance
 *       it cannot vouch for. Sealing is Matrix's own act. Evidence is
 *       <i>read</i> through the <b>server's</b> client, resolving an
 *       {@code [evidence/<id>#<row>]} citation.
 *   <li><b>No local validation.</b> {@link #validate(String)} posts the YAML
 *       and returns Matrix's own findings. A client carrying its own copy of
 *       the rules would drift from the service that enforces them, and the
 *       drift would surface as an asset that validates here and is refused
 *       there.
 *   <li><b>No SQL.</b> Nothing on this surface takes a statement. Queries are
 *       pre-declared contracts and views, executed by name.
 * </ul>
 *
 * <p>Nothing here retries, and that is a real trade-off rather than an
 * oversight. Matrix's mutating routes carry no idempotency-key contract, so a
 * blind re-send of an accepted {@code sync} queues a second one; meanwhile the
 * refusal class already tells a caller when trying again is meaningful
 * ({@link MatrixException#retryable()}). A built-in policy that ignored that
 * distinction would be worse than none, so the decision lands on never doing
 * twice what the caller asked once.
 *
 * <p>Instances are safe to share across threads; the underlying
 * {@code HttpClient} is.
 */
public final class MatrixClient implements AutoCloseable {

    private final HttpClient http;
    private final String base;
    private final Duration requestTimeout;
    private final String token;
    private final String uid;

    public MatrixClient(MatrixClientOptions options) {
        this.http = HttpClient.newBuilder().connectTimeout(options.connectTimeout()).build();
        this.base = options.endpoint().replaceAll("/+$", "");
        this.requestTimeout = options.requestTimeout();
        this.token = options.token();
        this.uid = options.uid();
    }

    public static MatrixClient of(String endpoint) {
        return new MatrixClient(MatrixClientOptions.of(endpoint));
    }

    public static MatrixClient of(String endpoint, String token) {
        return new MatrixClient(MatrixClientOptions.of(endpoint).withToken(token));
    }

    @Override
    public void close() {
        // Java 21's HttpClient is AutoCloseable: without this, each client
        // leaks a selector-manager thread and its pooled sockets until GC.
        http.close();
    }

    // -- meta -----------------------------------------------------------------

    /**
     * {@code GET /version} — including whether this Matrix's server agrees with
     * it about the contract. See {@link Version#lockstepOk()}.
     */
    public Version version() {
        return read("GET", "/version", null, Version.class);
    }

    /**
     * Liveness. Answers {@code false} rather than throwing: "is it up?" turned
     * into an exception is a question the caller has to wrap in a try just to
     * get a boolean back out of.
     */
    public boolean healthz() {
        try {
            exchange("GET", "/healthz", null, null, null);
            return true;
        } catch (MatrixException e) {
            return false;
        }
    }

    /**
     * Which sources are REGISTERED — see {@link HealthData}, which deliberately
     * is not a connectivity check.
     */
    public HealthData healthdata() {
        return read("GET", "/healthdata", null, HealthData.class);
    }

    // -- registry -------------------------------------------------------------

    /**
     * Apply one asset, kind-sniffed by Matrix from its own {@code kind:} line.
     * See {@link ApplyOutcome} for what {@code unchanged} means.
     */
    public ApplyOutcome apply(String yaml) {
        return writeYaml("POST", "/v1/assets", yaml, ApplyOutcome.class);
    }

    /** Matrix's own validators, run without applying anything. */
    public Validation validate(String yaml) {
        return writeYaml("POST", "/v1/assets/validate", yaml, Validation.class);
    }

    /**
     * List a registry. {@code kind} is the route segment: {@code datasources},
     * {@code contracts}, {@code mappings}, {@code metricviews},
     * {@code dataviews}.
     */
    public List<AssetSummary> listAssets(String kind) {
        return listAssets(kind, false);
    }

    /** {@code allVersions} asks for the history rather than the current version of each name. */
    public List<AssetSummary> listAssets(String kind, boolean allVersions) {
        // The parameter Matrix reads is `all_versions`. Spelling it `all` sends
        // something the service silently ignores, and a listing that is quietly
        // latest-only is indistinguishable from a registry with no history.
        String query = allVersions ? params(Map.of("all_versions", "true")) : null;
        JsonNode root = readTree("GET", "/v1/" + seg(kind), query);
        return convert(root.path("assets"), new TypeReference<List<AssetSummary>>() {});
    }

    /**
     * The applied YAML back, verbatim — the bytes Matrix stored, not a
     * re-serialisation of a parse. A round trip through the parsed form would
     * silently normalise an operator's file, and those stored bytes are what an
     * asset's identity is computed over.
     */
    public String getYaml(String kind, String name) {
        HttpResponse<byte[]> resp =
                exchange("GET", "/v1/" + seg(kind) + "/" + seg(name), null, null, null);
        return new String(resp.body(), StandardCharsets.UTF_8);
    }

    // -- sources --------------------------------------------------------------

    /**
     * What the source exposes, and what the configured role can actually do
     * there.
     *
     * <p>Left as a {@link JsonNode} deliberately: the role-posture report and
     * the table list are still moving, and mirroring an unsettled shape would
     * put a second normative copy of it in this client.
     */
    public JsonNode introspect(String source) {
        return readTree("POST", "/v1/datasources/" + seg(source) + "/introspect", null);
    }

    /**
     * Reachability now. An unreachable source is an ANSWER here, not an
     * exception — see {@link Probe}.
     */
    public Probe probe(String source) {
        return read("POST", "/v1/datasources/" + seg(source) + "/probe", null, Probe.class);
    }

    /** Enqueue a materialization pass (mode A). */
    public JobAccepted sync(String source) {
        return read("POST", "/v1/datasources/" + seg(source) + "/sync", null, JobAccepted.class);
    }

    // -- contracts and views --------------------------------------------------

    /**
     * Run a query contract's verified questions — its regression suite. The
     * call succeeding and the CONTRACT passing are different things: read
     * {@link VerifyOutcome#failed()}.
     */
    public VerifyOutcome verify(String contract) {
        return read("POST", "/v1/contracts/" + seg(contract) + "/verify", null, VerifyOutcome.class);
    }

    /**
     * The same for a metric view or a native data view, recording the
     * definition fingerprint the questions ran under.
     *
     * <p>A metric view first, a data view when there is none by that name: the
     * caller names the VIEW, not the route it happens to live on. Which of the
     * two kinds a view is, is an authoring detail, and making every call site
     * track it would mean every call site knows something it has no reason to.
     */
    public VerifyOutcome verifyView(String view) {
        try {
            return read("POST", "/v1/metricviews/" + seg(view) + "/verify", null, VerifyOutcome.class);
        } catch (MatrixException e) {
            if (e.status() == null || e.status() != 404) {
                throw e;
            }
            return read("POST", "/v1/dataviews/" + seg(view) + "/verify", null, VerifyOutcome.class);
        }
    }

    // -- reconcile ------------------------------------------------------------

    /** Enqueue a reconcile pass (mode C). */
    public JobAccepted reconcile(String mapping) {
        return read("POST", "/v1/mappings/" + seg(mapping) + "/run", null, JobAccepted.class);
    }

    /** Where a mapping stands against the promotion gates, promoted or not. */
    public PromotionStatus promotionStatus(String mapping) {
        return read("GET", "/v1/mappings/" + seg(mapping) + "/promotion", null, PromotionStatus.class);
    }

    /** The gate measurements over time, at whatever limit the service defaults to. */
    public GateHistory gateHistory(String mapping) {
        return gateHistory(mapping, 0);
    }

    /** A {@code limit} of zero or less leaves the count to the service. */
    public GateHistory gateHistory(String mapping, int limit) {
        String query = limit > 0 ? params(Map.of("limit", String.valueOf(limit))) : null;
        return read("GET", "/v1/mappings/" + seg(mapping) + "/gate-history", query, GateHistory.class);
    }

    /**
     * Let a mapping's claims reach the ledger.
     *
     * <p>The gates — identity precision, value conformance — are checked by
     * MATRIX at the decision, not here. A client that pre-checked them would be
     * a second opinion nobody audited, and it would disagree with the service
     * the moment a threshold moved.
     *
     * <p>{@code decisionId} is the operator's own record — a ticket, a change
     * number — and it is required, because a promotion nobody can trace to a
     * decision is a promotion nobody made.
     */
    public PromotionStatus promote(String mapping, String decisionId, String actor, String reason) {
        Map<String, String> body = new LinkedHashMap<>();
        body.put("decision_id", decisionId);
        if (actor != null) {
            body.put("actor", actor);
        }
        if (reason != null) {
            body.put("reason", reason);
        }
        return writeJson("POST", "/v1/mappings/" + seg(mapping) + "/promote", body, PromotionStatus.class);
    }

    /** As above, with no stated reason. */
    public PromotionStatus promote(String mapping, String decisionId, String actor) {
        return promote(mapping, decisionId, actor, null);
    }

    /**
     * Stop the writes. Nothing already proposed is touched — that is what
     * {@link #rollback} is for.
     */
    public PromotionStatus demote(String mapping, String decisionId) {
        return writeJson("POST", "/v1/mappings/" + seg(mapping) + "/demote",
                Map.of("decision_id", decisionId), PromotionStatus.class);
    }

    /**
     * Undo what a promoted mapping wrote — by SUPERSESSION, never by deletion.
     * History is not rewritten.
     */
    public RollbackOutcome rollback(String mapping, String decisionId) {
        return writeJson("POST", "/v1/mappings/" + seg(mapping) + "/rollback",
                Map.of("decision_id", decisionId), RollbackOutcome.class);
    }

    // -- audit ----------------------------------------------------------------

    /** The 50 most recent journal entries. */
    public List<JsonNode> journal() {
        return journal(50);
    }

    /**
     * Every operation, redacted by default: parameters and results never
     * appear, only what happened and how it ended.
     *
     * <p>Entries stay {@link JsonNode} because their shape varies by operation
     * kind. Flattening them into one record would either lose fields or invent
     * a union that nothing on the wire actually writes.
     */
    public List<JsonNode> journal(int limit) {
        JsonNode root = readTree("GET", "/v1/journal", params(Map.of("limit", String.valueOf(limit))));
        JsonNode entries = root.isArray() ? root : root.path("entries");
        List<JsonNode> out = new ArrayList<>();
        entries.forEach(out::add);
        return List.copyOf(out);
    }

    // -- plumbing -------------------------------------------------------------

    private static String seg(String s) {
        // Asset names are free-form enough that a stray '/' or '?' must not be
        // able to change the route shape. (URLEncoder is form-encoding: undo
        // the '+' it produces for a space.)
        return URLEncoder.encode(s, StandardCharsets.UTF_8).replace("+", "%20");
    }

    private static String params(Map<String, String> values) {
        StringBuilder sb = new StringBuilder();
        for (Map.Entry<String, String> e : values.entrySet()) {
            if (!sb.isEmpty()) {
                sb.append('&');
            }
            sb.append(URLEncoder.encode(e.getKey(), StandardCharsets.UTF_8))
                    .append('=')
                    .append(URLEncoder.encode(e.getValue(), StandardCharsets.UTF_8));
        }
        return sb.toString();
    }

    private <T> T read(String method, String path, String query, Class<T> type) {
        return decode(exchange(method, path, query, null, null), type);
    }

    private JsonNode readTree(String method, String path, String query) {
        return decode(exchange(method, path, query, null, null), JsonNode.class);
    }

    private <T> T writeYaml(String method, String path, String yaml, Class<T> type) {
        BodyPublisher body = BodyPublishers.ofString(yaml, StandardCharsets.UTF_8);
        return decode(exchange(method, path, null, body, "text/yaml"), type);
    }

    private <T> T writeJson(String method, String path, Object body, Class<T> type) {
        byte[] bytes;
        try {
            bytes = MatrixJson.MAPPER.writeValueAsBytes(body);
        } catch (IOException e) {
            throw new MatrixException(
                    "unserializable request body: " + e.getMessage(), null, "invalid", null, null);
        }
        return decode(
                exchange(method, path, null, BodyPublishers.ofByteArray(bytes), "application/json"),
                type);
    }

    private HttpResponse<byte[]> exchange(
            String method, String path, String query, BodyPublisher body, String contentType) {
        URI uri = URI.create(base + path + (query == null || query.isEmpty() ? "" : "?" + query));
        HttpRequest.Builder b = HttpRequest.newBuilder(uri).timeout(requestTimeout);
        if (token != null) {
            b.header("authorization", "Bearer " + token);
        }
        if (uid != null) {
            b.header("x-munarium-uid", uid);
        }
        if (contentType != null) {
            b.header("content-type", contentType);
        }
        b.method(method, body == null ? BodyPublishers.noBody() : body);
        HttpResponse<byte[]> resp;
        try {
            resp = http.send(b.build(), BodyHandlers.ofByteArray());
        } catch (IOException e) {
            throw MatrixException.transportFailure(e);
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
            throw MatrixException.transportFailure(e);
        }
        if (resp.statusCode() >= 200 && resp.statusCode() < 300) {
            return resp;
        }
        throw MatrixException.from(resp.statusCode(), resp.body());
    }

    private static <T> T decode(HttpResponse<byte[]> resp, Class<T> type) {
        try {
            return MatrixJson.MAPPER.readValue(resp.body(), type);
        } catch (IOException e) {
            throw new MatrixException(
                    "undecodable success body: " + e.getMessage(), resp.statusCode(), null, null, null);
        }
    }

    private static <T> T convert(JsonNode node, TypeReference<T> type) {
        JsonNode source = node;
        if (source == null || source.isMissingNode() || source.isNull()) {
            source = MatrixJson.MAPPER.createArrayNode();
        }
        return MatrixJson.MAPPER.convertValue(source, type);
    }
}

File: clients/matrix-dotnet/src/Ioka.Munarium.Matrix.Client/MatrixClient.cs
// SPDX-License-Identifier: Apache-2.0
// The client. One transport (REST), one request pipeline, and a surface that
// is small because Matrix's whole job is small: register assets, run the three
// modes, read what happened.
//
// Async only, and not as a style preference. A synchronous wrapper over
// HttpClient is the classic way to deadlock an application: .Result on a
// captured synchronization context blocks the thread the continuation needs,
// and the failure shows up in production under load rather than in a test. A
// client that offers a sync overload will have that overload called, so the
// only safe number of them is zero. The Python client ships sync and async
// because Python's sync HTTP is a genuinely separate, safe implementation;
// .NET has no such thing to offer.

using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization.Metadata;

namespace Ioka.Munarium.Matrix.Client;

/// <summary>Connection options.</summary>
public sealed record MatrixClientOptions
{
    /// <summary>Matrix's REST base URL (http://host:8180, or :443 behind a
    /// gateway). Not the gRPC endpoint — see <see cref="MatrixClient"/>.</summary>
    public required string Endpoint { get; init; }

    /// <summary>Bearer token. Matrix gates routes by the token's role, so a
    /// read-only token 403s on apply and a non-mgmt token 403s on the
    /// journal; that is the deployment's answer, not this client's.</summary>
    public string? Token { get; init; }

    /// <summary>The acting operator id, sent as <c>X-Munarium-Uid</c>.
    /// Matrix itself does not read it — the munarium-server does, on every
    /// /v1 request — but sending it keeps one identity across both journals
    /// when the same operator drives them.</summary>
    public string? Uid { get; init; }

    /// <summary>Per-request deadline. Long enough for a verify that runs a
    /// contract's whole suite against a cold warehouse, short enough that a
    /// wedged call is not forever.</summary>
    public TimeSpan Timeout { get; init; } = TimeSpan.FromSeconds(30);
}

/// <summary>The .NET client for Munarium Matrix, the structured-evidence
/// plane.</summary>
/// <remarks>
/// <para>
/// <b>REST only — there is no gRPC transport here.</b> Matrix's gRPC plane
/// serves <c>Execute</c> alone, and <c>Execute</c> is service-to-service: the
/// munarium-server calls it while answering a turn, so that the evidence an
/// answer cites is sealed by the same process that ran the query. An
/// application is on the other side of the server, asking for an answer. If
/// that ever changes, this client grows a transport rather than acquiring a
/// sibling.
/// </para>
/// <para>
/// Three absences, each a decision rather than a missing feature.
/// <b>No sealing:</b> a manifest is a statement about work the <i>sealer</i>
/// did, so an SDK offering it would invite an application to assert provenance
/// it cannot vouch for. Evidence is <i>read</i> through the server's client,
/// resolving an <c>[evidence/&lt;id&gt;#&lt;row&gt;]</c> citation.
/// <b>No local validation:</b> <see cref="ValidateAsync"/> posts the YAML and
/// returns Matrix's own findings, because a client carrying its own copy of
/// the rules would drift from the service that enforces them, and the drift
/// would surface as an asset that validates here and is refused there.
/// <b>No SQL:</b> nothing on this surface takes a statement; queries are
/// pre-declared contracts and views, executed by name.
/// </para>
/// </remarks>
public sealed class MatrixClient : IDisposable
{
    /// <summary>The Matrix and munarium-server version this client is in
    /// lockstep with. A wire-surface change bumps all three together.</summary>
    public const string TargetVersion = "1.0.0";

    private readonly HttpClient _http;
    private readonly bool _ownsHttp;
    private readonly string _base;
    private readonly TimeSpan _timeout;
    private readonly AuthenticationHeaderValue? _auth;
    private readonly string? _uid;

    /// <summary>Construct against an endpoint. Pass an
    /// <paramref name="httpClient"/> from IHttpClientFactory when you manage
    /// pooling yourself — a caller-supplied client is never mutated, since
    /// auth rides each request.</summary>
    public MatrixClient(MatrixClientOptions options, HttpClient? httpClient = null)
    {
        ArgumentNullException.ThrowIfNull(options);
        _ownsHttp = httpClient is null;
        _http = httpClient ?? new HttpClient(new SocketsHttpHandler
        {
            ConnectTimeout = TimeSpan.FromSeconds(5),
            // The client is explicitly long-lived; without a lifetime the pool
            // pins retired IPs across a DNS cutover forever.
            PooledConnectionLifetime = TimeSpan.FromMinutes(5),
        })
        {
            // The per-request deadline rides the attempt token instead, so a
            // deadline expiry can be typed as a transport failure rather than
            // arriving as a bare TaskCanceledException.
            Timeout = System.Threading.Timeout.InfiniteTimeSpan,
        };
        _base = options.Endpoint.TrimEnd('/');
        _timeout = options.Timeout;
        _auth = options.Token is null
            ? null
            : new AuthenticationHeaderValue("Bearer", options.Token);
        _uid = options.Uid;
    }

    /// <summary>Convenience over <see cref="MatrixClientOptions"/>.</summary>
    public MatrixClient(string endpoint, string? token = null, string? uid = null)
        : this(new MatrixClientOptions { Endpoint = endpoint, Token = token, Uid = uid })
    {
    }

    public void Dispose()
    {
        if (_ownsHttp) _http.Dispose();
    }

    // -- meta ---------------------------------------------------------------

    /// <summary>Matrix's version, its contract version, the role it serves,
    /// and whether it agrees with the server it seals into.</summary>
    public Task<MatrixVersion> VersionAsync(CancellationToken ct = default) =>
        SendAsync(HttpMethod.Get, "/version", null, MatrixJsonContext.Default.MatrixVersion, ct);

    /// <summary>Liveness. A refusal is an ANSWER here, so this returns false
    /// rather than throwing — a health check that throws forces every caller
    /// to write the try/catch this one has already written.</summary>
    public async Task<bool> HealthzAsync(CancellationToken ct = default)
    {
        try
        {
            var body = await SendAsync(
                HttpMethod.Get, "/healthz", null,
                MatrixJsonContext.Default.JsonElement, ct).ConfigureAwait(false);
            return body.ValueKind == JsonValueKind.Object
                && body.TryGetProperty("ok", out var ok)
                && ok.ValueKind == JsonValueKind.True;
        }
        catch (MatrixException)
        {
            return false;
        }
    }

    /// <summary>Registration health per source. Registration, NOT
    /// connectivity: probing every source on a health call would make a
    /// health endpoint an outbound-traffic amplifier.</summary>
    public Task<JsonElement> HealthdataAsync(CancellationToken ct = default) =>
        SendAsync(HttpMethod.Get, "/healthdata", null, MatrixJsonContext.Default.JsonElement, ct);

    // -- registry -----------------------------------------------------------

    /// <summary>Apply one asset, kind-sniffed by Matrix from its
    /// <c>kind:</c> line.</summary>
    /// <remarks>Re-applying identical bytes is
    /// <see cref="ApplyOutcome.Unchanged"/>, not an error: that is ordinary
    /// GitOps. The same version with DIFFERENT bytes is refused, because a
    /// version is provenance — sealed evidence cites it.</remarks>
    public Task<ApplyOutcome> ApplyAsync(string yaml, CancellationToken ct = default) =>
        SendAsync(
            HttpMethod.Post, "/v1/assets", Yaml(yaml),
            MatrixJsonContext.Default.ApplyOutcome, ct);

    /// <summary>Matrix's own findings, from the same validators
    /// <c>mxctl validate</c> uses, without applying anything.</summary>
    public Task<ValidationOutcome> ValidateAsync(string yaml, CancellationToken ct = default) =>
        SendAsync(
            HttpMethod.Post, "/v1/assets/validate", Yaml(yaml),
            MatrixJsonContext.Default.ValidationOutcome, ct);

    /// <summary>List applied assets of one kind.</summary>
    /// <param name="kind">A route segment: datasources, contracts, mappings,
    /// metricviews, dataviews.</param>
    /// <param name="allVersions">Every version rather than the latest.</param>
    /// <param name="ct">Cancellation.</param>
    public async Task<IReadOnlyList<AssetSummary>> ListAssetsAsync(
        string kind, bool allVersions = false, CancellationToken ct = default)
    {
        // The parameter is `all_versions`. Anything else is ignored in silence
        // by the service's deserializer, which is the worst possible failure
        // for a flag: the call succeeds and answers the other question.
        var query = allVersions ? "?all_versions=true" : "";
        var page = await SendAsync(
            HttpMethod.Get, $"/v1/{Seg(kind)}{query}", null,
            MatrixJsonContext.Default.AssetListResponse, ct).ConfigureAwait(false);
        return page.Assets;
    }

    /// <summary>The applied YAML, verbatim — the bytes Matrix stored, not a
    /// re-serialisation of a parse.</summary>
    public Task<string> GetYamlAsync(string kind, string name, CancellationToken ct = default) =>
        SendTextAsync(HttpMethod.Get, $"/v1/{Seg(kind)}/{Seg(name)}", null, ct);

    // -- operations ---------------------------------------------------------

    /// <summary>What the configured role can actually see, and what the
    /// schema holds — with a draft contract the operator edits, never applied
    /// automatically.</summary>
    public Task<JsonElement> IntrospectAsync(string source, CancellationToken ct = default) =>
        SendAsync(
            HttpMethod.Post, $"/v1/datasources/{Seg(source)}/introspect", null,
            MatrixJsonContext.Default.JsonElement, ct);

    /// <summary>Reachability now. A refusal is an ANSWER here —
    /// <c>reachable: false</c> with a typed reason — not an exception.</summary>
    public Task<JsonElement> ProbeAsync(string source, CancellationToken ct = default) =>
        SendAsync(
            HttpMethod.Post, $"/v1/datasources/{Seg(source)}/probe", null,
            MatrixJsonContext.Default.JsonElement, ct);

    /// <summary>Queue a materialization (mode A), one job per authorization
    /// class — a collection carries exactly one class, so a multi-class
    /// source needs one run each.</summary>
    public Task<JobAccepted> SyncAsync(string source, CancellationToken ct = default) =>
        SendAsync(
            HttpMethod.Post, $"/v1/datasources/{Seg(source)}/sync", null,
            MatrixJsonContext.Default.JobAccepted, ct);

    /// <summary>Run a query contract's verified questions — its regression
    /// suite.</summary>
    /// <remarks>The call succeeding and the CONTRACT passing are different
    /// things: check <see cref="VerifyOutcome.Failed"/>. <c>mxctl</c> exits 3
    /// on a non-zero <c>failed</c> for exactly this reason, so CI can tell a
    /// broken contract from a broken command.</remarks>
    public Task<VerifyOutcome> VerifyAsync(string contract, CancellationToken ct = default) =>
        SendAsync(
            HttpMethod.Post, $"/v1/contracts/{Seg(contract)}/verify", null,
            MatrixJsonContext.Default.VerifyOutcome, ct);

    /// <summary>The same for a metric view or a native data view, recording
    /// the definition fingerprint the questions ran under.</summary>
    /// <remarks>
    /// <para>
    /// A metric view first, a data view when there is none by that name — the
    /// caller names the view, not the route it happens to live on.
    /// </para>
    /// <para>
    /// The fallback fires on two different answers, because Matrix spells
    /// "no such asset" two ways: a bare 404 when the store misses, and a
    /// <c>not_covered</c> refusal (HTTP 422) when the miss goes through the
    /// asset loader — which is the path <c>/v1/metricviews/{name}/verify</c>
    /// actually takes. Keying on 404 alone reads correctly and never fires,
    /// so every data view would be reported as an unknown metric view.
    /// </para>
    /// <para>
    /// When the fallback ALSO says no such asset, the metric-view error is
    /// what surfaces: the view exists nowhere, and the first answer is the
    /// one the caller asked for. When the fallback fails for any other reason
    /// the data view exists and something else went wrong, so that error
    /// surfaces instead — reporting "no such metric view" there would hide a
    /// real fault behind a naming detail.
    /// </para>
    /// </remarks>
    public async Task<VerifyOutcome> VerifyViewAsync(string view, CancellationToken ct = default)
    {
        MatrixException first;
        try
        {
            return await SendAsync(
                HttpMethod.Post, $"/v1/metricviews/{Seg(view)}/verify", null,
                MatrixJsonContext.Default.VerifyOutcome, ct).ConfigureAwait(false);
        }
        catch (MatrixException e) when (e.IsNoSuchAsset)
        {
            first = e;
        }

        try
        {
            return await SendAsync(
                HttpMethod.Post, $"/v1/dataviews/{Seg(view)}/verify", null,
                MatrixJsonContext.Default.VerifyOutcome, ct).ConfigureAwait(false);
        }
        catch (MatrixException e) when (e.IsNoSuchAsset)
        {
            throw first;
        }
    }

    /// <summary>Queue a reconcile pass (mode C).</summary>
    public Task<JobAccepted> ReconcileAsync(string mapping, CancellationToken ct = default) =>
        SendAsync(
            HttpMethod.Post, $"/v1/mappings/{Seg(mapping)}/run", null,
            MatrixJsonContext.Default.JobAccepted, ct);

    // -- promotion ----------------------------------------------------------

    /// <summary>Whether a mapping may write canon, the gate numbers from its
    /// latest completed run, and the state of the most recent pass.</summary>
    public Task<PromotionStatus> PromotionStatusAsync(
        string mapping, CancellationToken ct = default) =>
        SendAsync(
            HttpMethod.Get, $"/v1/mappings/{Seg(mapping)}/promotion", null,
            MatrixJsonContext.Default.PromotionStatus, ct);

    /// <summary>The promotion gates over time, newest first, each scored
    /// against the thresholds in force RIGHT NOW.</summary>
    /// <remarks>Which is the point: lowering a threshold and re-reading this
    /// says exactly which past runs the new number would have admitted,
    /// before anything is promoted under it.</remarks>
    public Task<JsonElement> GateHistoryAsync(
        string mapping, int? limit = null, CancellationToken ct = default) =>
        SendAsync(
            HttpMethod.Get,
            $"/v1/mappings/{Seg(mapping)}/gate-history{(limit is null ? "" : $"?limit={limit}")}",
            null, MatrixJsonContext.Default.JsonElement, ct);

    /// <summary>Let a mapping's claims reach the ledger.</summary>
    /// <remarks>
    /// The gates (identity precision, value conformance) are checked by
    /// MATRIX at the decision, not here: a client that pre-checked them would
    /// be a second opinion nobody audited. <paramref name="decisionId"/> is
    /// the operator's record — a ticket, a change number — and it is required
    /// because a promotion nobody can trace to a decision is a promotion
    /// nobody made.
    /// </remarks>
    public Task<PromotionStatus> PromoteAsync(
        string mapping, string decisionId, string actor, string? reason = null,
        CancellationToken ct = default) =>
        SendAsync(
            HttpMethod.Post, $"/v1/mappings/{Seg(mapping)}/promote",
            Json(new PromoteBody { DecisionId = decisionId, Actor = actor, Reason = reason },
                MatrixJsonContext.Default.PromoteBody),
            MatrixJsonContext.Default.PromotionStatus, ct);

    /// <summary>Stop the writes, effective on the next reconcile poll.
    /// Nothing already proposed is touched — that is what
    /// <see cref="RollbackAsync"/> is for.</summary>
    public Task<PromotionStatus> DemoteAsync(
        string mapping, string decisionId, CancellationToken ct = default) =>
        SendAsync(
            HttpMethod.Post, $"/v1/mappings/{Seg(mapping)}/demote",
            Json(new DecisionBody { DecisionId = decisionId },
                MatrixJsonContext.Default.DecisionBody),
            MatrixJsonContext.Default.PromotionStatus, ct);

    /// <summary>Undo what a promoted mapping wrote — by SUPERSESSION, never
    /// by deletion. History is not rewritten, and the correcting claims carry
    /// <c>origin.kind = "rollback"</c> so a reviewer sees both moves.</summary>
    public Task<JsonElement> RollbackAsync(
        string mapping, string decisionId, CancellationToken ct = default) =>
        SendAsync(
            HttpMethod.Post, $"/v1/mappings/{Seg(mapping)}/rollback",
            Json(new DecisionBody { DecisionId = decisionId },
                MatrixJsonContext.Default.DecisionBody),
            MatrixJsonContext.Default.JsonElement, ct);

    // -- journal ------------------------------------------------------------

    /// <summary>Every operation, redacted by default: parameters and results
    /// never appear, only what happened and how it ended. Management role
    /// only — a journal readable by the tokens it audits is not an audit
    /// log.</summary>
    public async Task<IReadOnlyList<JsonElement>> JournalAsync(
        int limit = 50, CancellationToken ct = default)
    {
        var page = await SendAsync(
            HttpMethod.Get, $"/v1/journal?limit={limit}", null,
            MatrixJsonContext.Default.JournalListResponse, ct).ConfigureAwait(false);
        return page.Entries;
    }

    // -- the request pipeline -----------------------------------------------

    /// <summary>Percent-encode a path segment. Asset names are free-form
    /// enough that a raw '/' or '?' must not change the route shape.</summary>
    private static string Seg(string s) => Uri.EscapeDataString(s);

    private static Func<HttpContent> Yaml(string yaml) =>
        () => new StringContent(yaml, Encoding.UTF8, "text/yaml");

    private static Func<HttpContent> Json<TBody>(TBody body, JsonTypeInfo<TBody> type)
    {
        // Serialize once; the content object itself is single-use.
        var bytes = JsonSerializer.SerializeToUtf8Bytes(body, type);
        return () =>
        {
            var content = new ByteArrayContent(bytes);
            content.Headers.ContentType = new MediaTypeHeaderValue("application/json");
            return content;
        };
    }

    /// <summary>THE one request path: build, send under the deadline, decode a
    /// non-success body through the refusal registry, decode a success body
    /// into the typed model.</summary>
    /// <remarks>
    /// There is deliberately no retry loop. Matrix's refusals already say
    /// whether retrying could help (<see cref="MatrixException.Retryable"/>),
    /// and the two classes that could — unavailable and exhausted — want
    /// pacing the caller owns: an exhausted budget refusal carries the wait
    /// the service asked for, and burning it down inside a client would spend
    /// the caller's budget on its behalf.
    /// </remarks>
    private async Task<HttpResponseMessage> SendRawAsync(
        HttpMethod method, string pathAndQuery, Func<HttpContent>? content,
        CancellationTokenSource attempt, CancellationToken ct)
    {
        using var request = new HttpRequestMessage(method, _base + pathAndQuery);
        request.Headers.Authorization = _auth;
        if (_uid is not null) request.Headers.Add("X-Munarium-Uid", _uid);
        if (content is not null) request.Content = content();
        try
        {
            return await _http
                .SendAsync(request, HttpCompletionOption.ResponseHeadersRead, attempt.Token)
                .ConfigureAwait(false);
        }
        catch (OperationCanceledException) when (ct.IsCancellationRequested)
        {
            throw; // caller cancellation is not a transport failure
        }
        catch (Exception e)
            when (e is HttpRequestException or OperationCanceledException or IOException)
        {
            throw Errors.FromTransport(e);
        }
    }

    private async Task<string> ReadBodyAsync(
        HttpResponseMessage resp, CancellationTokenSource attempt, CancellationToken ct)
    {
        try
        {
            return await resp.Content.ReadAsStringAsync(attempt.Token).ConfigureAwait(false);
        }
        catch (OperationCanceledException) when (ct.IsCancellationRequested)
        {
            throw;
        }
        catch (Exception e)
            when (e is HttpRequestException or OperationCanceledException or IOException)
        {
            throw Errors.FromTransport(e);
        }
    }

    private async Task<T> SendAsync<T>(
        HttpMethod method, string pathAndQuery, Func<HttpContent>? content,
        JsonTypeInfo<T> responseType, CancellationToken ct)
    {
        var body = await SendTextAsync(method, pathAndQuery, content, ct).ConfigureAwait(false);
        try
        {
            return JsonSerializer.Deserialize(body, responseType)
                ?? throw new MatrixException("matrix answered an empty body");
        }
        catch (JsonException e)
        {
            throw new MatrixException($"undecodable success body: {e.Message}", inner: e);
        }
    }

    private async Task<string> SendTextAsync(
        HttpMethod method, string pathAndQuery, Func<HttpContent>? content, CancellationToken ct)
    {
        using var attempt = CancellationTokenSource.CreateLinkedTokenSource(ct);
        attempt.CancelAfter(_timeout);
        using var resp = await SendRawAsync(method, pathAndQuery, content, attempt, ct)
            .ConfigureAwait(false);
        var body = await ReadBodyAsync(resp, attempt, ct).ConfigureAwait(false);
        if (!resp.IsSuccessStatusCode)
        {
            throw Errors.FromProblem((int)resp.StatusCode, body);
        }
        return body;
    }
}
